cilium/CFP-33716-distributed-packet-tracing.md

```diff
@@ -0,0 +1,424 @@
+- Tx time: 1.34734s | Tx bytes/s: 32657.01 | Tx pkts/s: 742.20
+- Rx time: 1.34779s | Rx bytes/s: 32646.04 | Rx pkts/s: 741.96
+- Nping done: 1 IP address pinged in 1.45 seconds
+```
+
+## With tracing feature enabled:
+
+```sh
+Baseline (without IP Options):
+- Max rtt: 0.872ms | Min rtt: 0.046ms | Avg rtt: 0.046ms
+- Raw packets sent: 1000 (40.000KB) | Rcvd: 1000 (44.000KB) | Lost: 0 (0.00%)
+- Tx time: 1.33641s | Tx bytes/s: 29931.00 | Tx pkts/s: 748.28
+- Rx time: 1.33649s | Rx bytes/s: 32922.03 | Rx pkts/s: 748.23
+- Nping done: 1 IP address pinged in 1.45 seconds
+
+Test (with IP Options):
+- Max rtt: 0.719ms | Min rtt: 0.048ms | Avg rtt: 0.048ms
+- Raw packets sent: 1000 (44.000KB) | Rcvd: 1000 (44.000KB) | Lost: 0 (0.00%)
+- Tx time: 1.34727s | Tx bytes/s: 32658.61 | Tx pkts/s: 742.24
+- Rx time: 1.34733s | Rx bytes/s: 32657.11 | Rx pkts/s: 742.21
+- Nping done: 1 IP address pinged in 1.46 seconds
+```
+
+The additional delay introduced by enabling the tracing feature is relatively small. When the feature is disabled, the average RTT (round-trip time) remains the same (0.046ms).
+
+When the tracing feature is enabled, the average RTT increases by 0.002ms when IP Options are used (0.046ms vs 0.048ms). Overall, the performance impact of enabling the tracing feature and using IP Options is minimal, making it a feasible solution for environments requiring detailed packet tracing without significant degradation in network performance. Moreover, as this is a debugging feature, in environments where this feature is disabled, there is no performance loss.
+
+## Future Milestones
+
+### Deferred Milestone 1: Use conntrack to restore trace ID for reply packets
+- When a packet includes a trace ID, the conntrack entry should record the ID along with the connection information.
+- For subsequent packets – if a trace ID is not present in the IP Options – use the track ID from conntrack to restore the internal state.
+- Note this will not modify the IP option header.
+
+### Deferred Milestone 2: Tag packets
+- **Audience:** Proficient Cilium user with knowledge of cilium-dbg / pwru tools.
+- Provide a `cilium-dbg` subcommand for a user to define what traffic should be traced.
+- Prepend tag program to container lxc interface.
+- **Optional:** Provide a flag to optionally remove the tag before egressing Node NIC, addressing targets that do not support IP options.
+- **Challenge:** This affects the available MTU.
+  - [Preferred] Only apply to packets that have the length available (e.g., packets with no data; SYN w/o fast open, RST, etc.).
+  - Modify the MTU/route MTU of the container (i.e., subtracting the IP option + trace ID lengths).
+
+### Deferred Milestone 3: Tag packets via CR
+- **Audience:** Network specialist with knowledge of network topology.
+- Develop a K8s API where the user specifies traffic to tag using common selectors: 4-tuple, pod identity, etc.
+
+## Appendix
```

Comment:
It might be worth investigating or declaring how we would expect this to work with some of Cilium's other features:
- L7 policies/Envoy
- Egress gateway
-  IPSec/Wireguard.

It's okay to not know, but we should at least explicitly mention it's unsupported with certain other features that we could expect it to not work with, so we can document that if the CFP is accepted and implemented.

Reply:
There is now a section addressing this. More investigation is ongoing, but this is what we expect based on current context/information. Please take a look and feel free to ask follow-ups/any clarifications.